### PR TITLE
Allow index & cluster default refresh interval setting value to be -1

### LINKED ISSUE #11410: [BUG] Allow index & cluster default refresh interval to be set as -1 (no periodic background refresh)
**Describe the bug**
We have added a setting `cluster.minimum.index.refresh_interval` which is validated against index creation, updation and index template creation (from api, or snapshot restore). Currently there is a bug that disallow setting index setting as -1 if the `cluster.minimum.index.refresh_interval` setting is not -1. Refresh interval of -1 corresponds to no background periodic refreshes and this should be allowed to be set for both index setting `index.refresh_interval` as well as the `cluster.default.index.refresh_interval` cluster setting.

**To Reproduce**
Update the minimum setting to zero or higher time. After that try creating/updating an index. This will fail with validation failure.

**Expected behavior**
The index setting `index.refresh_interval` and cluster setting `cluster.default.index.refresh_interval` regardless of the `cluster.minimum.index.refresh_interval` setting's value.

### PR DESCRIPTION
**Pull request by ashking94**
### Description
This PR fixes the bug that does not allow the index setting `index.refresh_interval ` and cluster setting `cluster.default.index.refresh_interval ` to be set to -1.  More details on https://github.com/opensearch-project/OpenSearch/issues/11410. 

### Related Issues
Resolves #11410

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/IndicesService.java [hunk 1]
```diff
@@ -282,8 +282,8 @@ public class IndicesService extends AbstractLifecycleComponent
      */
     public static final Setting<TimeValue> CLUSTER_MINIMUM_INDEX_REFRESH_INTERVAL_SETTING = Setting.timeSetting(
         "cluster.minimum.index.refresh_interval",
-        IndexSettings.MINIMUM_REFRESH_INTERVAL,
-        IndexSettings.MINIMUM_REFRESH_INTERVAL,
+        TimeValue.ZERO,
```
Comment (kkmr) [inline server/src/main/java/org/opensearch/indices/IndicesService.java:285]: Any reason why we need to change this?

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -1497,10 +1497,16 @@ public static void validateTranslogRetentionSettings(Settings indexSettings) {
...
     public static void validateRefreshIntervalSettings(Settings requestSettings, ClusterSettings clusterSettings) {
-        if (IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.exists(requestSettings) == false) {
+        if (IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.exists(requestSettings) == false
+            || requestSettings.get(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey()) == null) {
             return;
         }
         TimeValue requestRefreshInterval = IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.get(requestSettings);
+        // If the refresh interval supplied is -1, we allow the index to be created because -1 means no periodic refresh.
+        if (requestRefreshInterval.millis() < 0) {
+            assert requestRefreshInterval.millis() == -1;
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:1507]: Nitpick, but I'd remove this. The setting itself defines the legal range and unless you need specific logic built around a constraint related to the `-1` value then its probably better to rely on a single place to enforce min/max.

**Review by andrross — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/cluster/metadata/ClusterIndexRefreshIntervalIT.java [hunk 3]
```diff
@@ -235,33 +235,20 @@ public void testDefaultRefreshIntervalWithUpdateClusterAndIndexSettings() throws
...
-            IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> createIndex(INDEX_NAME, settings));
-            assertEquals(
-                "invalid index.refresh_interval [-1]: cannot be smaller than cluster.minimum.index.refresh_interval ["
-                    + getMinRefreshIntervalForRefreshDisabled()
-                    + "]",
-                exception.getMessage()
-            );
-        }
+        createIndex(INDEX_NAME, settings);
+        ensureYellowAndNoInitializingShards(INDEX_NAME);
```
[the remark this answers is not in the document]
Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/cluster/metadata/ClusterIndexRefreshIntervalIT.java:245]: Not really, it is a redundant check. This was added earlier. In this PR, we are just removed the if condition and let the if part run always.

File: server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java (hunk 2 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/cluster/metadata/MetadataCreateIndexService.java:1507]: I had added this for just verifying that the only legal value if refresh interval is less than 0 is -1. I was anyways going to remove this assertion and have a stricter check of -1 value.

**Comment by ashking94** [comment]
> ❌ Gradle check result for [fb11796](https://github.com/opensearch-project/OpenSearch/commit/fb1179675d800689f1aa045c9f3d9b42b69f91d1): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/30685/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test - https://github.com/opensearch-project/OpenSearch/issues/9891

**Review by kkmr — APPROVED** [review]

File: server/src/main/java/org/opensearch/indices/IndicesService.java [hunk 4]
```diff
@@ -1993,6 +1993,9 @@ public Iterator<Setting<?>> settings() {
      * @param defaultRefreshInterval value of cluster default index refresh interval setting
      */
     private static void validateRefreshIntervalSettings(TimeValue minimumRefreshInterval, TimeValue defaultRefreshInterval) {
+        if (defaultRefreshInterval.millis() < 0) {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/indices/IndicesService.java:1996]: Is this to allow setting cluster level default refresh interval to -1?

**Review by sachinpkale — APPROVED** [review]

File: server/src/main/java/org/opensearch/indices/IndicesService.java (hunk 4 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:1996]: Yes @sachinpkale

**Comment by ashking94** [comment]
> Will prefer a more predictable value rather than disabling a mode meant to perform constant work in the system(segment creation and replication)

Can you elaborate more on this, @Bukhtawar?

**Comment by gbbafna** [comment]
@astute-decipher  : can you pick this one and rebase it ?

**Comment by ashking94** [comment]
@gbbafna  let me do that for you. @astute-decipher might not have the access for it.

**Review by gbbafna — APPROVED** [review]

**Pull request merged**